### PR TITLE
Refactor MC part of AOD producers

### PR DESCRIPTION
**Pull request by cholmcc**
The MC part of the AOD producer workflow
`o2::aodproducer::AODProducerWorkflowDPL` and
`o2::aodmcproducer::AODMcProducerWorkflowDPL` is refactored to use
functions from namespace `o2::aodmchelpers`.   The helpers are

- `updateMCCollisions` which takes in the `MCEventHeader` and writes to
  the `o2::aod::McCollisions` table.
- `updateHepMCXSection` which takes the `MCEventHeader` and writes to
  the `o2::aodHepMCSections` table.  This uses the predefined key
  constants as defined in `o2::dataformats::MCInfoKeys`
- `updateHepMCPdfInfo` similar to `updateHepMCXSection` above
- `updateHepMCHeavyIon` similar to `updateHepMCXSection` above
- `updateParticle` uses information from an `o2::MCTrack` and writes it
  to the `o2::aod::McParticles` table
- `updateParticles` loops over `o2::MCTrack` objects and calls
  `updateParticle`.

These functions, in particular `updateHepMC...` uses the functions
- `hasKeys` which checks if the `MCEventHeader` has any or all of the
  keys queried.
- `getEventInfo` gets auxiliary information from the `MCEventHeader` or
  a default value.

For the `o2::aod::HepMC...` tables: Depending on the policy parameter
passed to the `updateHepMC...` functions, these tables may or may not be
updated.
- If the policy is `HepMCUpdate::never` then the tables are never
  updated.
- If the policy is `HepMCUpdate::always` then the tables are _always_
  updated, possibly with default values.
- If the policy is `HepMCUpdate::anyKey` (default) or `HepMCUpdate::allKeys`, then
  the decision of what to do is taken on the first event seen.
  - If the policy is `HepMCUpdate::anyKey`, then if _any_ of the needed
    keys are present, then updating will be enabled for this and _all_
    subsequent events.
  - If the policy is `HepMCUpdate::allKeys`, then if _all_ of the needed
    keys are present, then updating will be enabled for this and _all_
    subsequent events.
  
  Note that the availability of keys is _not_ checked after the first
  event.

  That means, if the criteria isn't met on the first event, then
  the tables will _never_ be update (as if the policy was
  `HepMCUpdate::never`).

  On the other hand, if the criteria was met, than the tables _will_ be
  update an all events (as if the policy was `HepMCUpdate::always`).

Note the slightly tricky template `TableCursor` which allows us to
define a type that corresponds to a table cursor (which is really a
lambda).   This template could be moved to `AnalysisDataFormats.h` or
the like.

The applications `o2-aod-producer-workflow` and
`o2-aod-mc-producer-workflow` have been updated (via their respective
implementation classes) to use these tools, thus unifying how the MC
information is propagated to AODs.

The utility `o2-sim-mctracks-to-aod` (`run/o2sim_mctracks_to_aod.cxx`)
has _also_ been updated to use these common tools.

Both `o2-aod-mc-producer-workflow` and `o2-sim-mctracks-to-aod` has been
tested extensively.  `o2-aod-producer-workflow` has _not_ been tested
since it is not clear to me how to set-up such a test with limited
resources (I tried the `prodtest/full_system_test.sh` but it failed in unrelated places - can't remember exactly where - due to some missing digits or the like).  However, since the changes _only_ effect the MC part, and
that part is now common between the two `o2-aod-{,mc-}producer-workflow`
applications, I believe there is no reason to think that it wouldn't
work.

**Note** Currently the helper functions, implemented in `Detectors/AOD/src/AODMcProducerHelpers.cxx` are directly compiled into  `o2-aod-producer-workflow`, `o2-aod-mc-producer-workflow`, and  `o2-sim-mctracks-to-aod`.  Ideally the code in `Detectors/AOD/src/AODMcProducerHelpers.cxx`  should probably live in a shared library linked to by these three applications.  However, at the moment there's no clear - at least to me - library to put that code into. 

BTW, the list of commits looks really long (60+ commits), but that's only because this MR builds upon two previous MRs that have been merged.  The number of files changed is small (12), though some of the changes are relatively large.  This is because large fractions have been refactored out and in to `Detectors/AOD/src/AODMcProducerHelpers.cxx`. 

Yours,
_Christian_

### CODE REVIEW

**Comment by cholmcc** [comment]
Post mortem on checks 

- `build/AliceO2/O2/o2/macOS`: ALICE fork of Apache **arrow** does not protect against `PREALLOCATE` already defined in system headers.  Not a problem from this MR but a problem in ALICE fork of **arrow** - **False Negative**  
- `build/AliceO2/O2/o2/macOS-arm`: Same as above - **False Negative**
- `build/O2/o2-cs8`: A QC test `multinode-test.json` fails due to time-out.  Seemingly unrelated to this MR - **False Negative**
- `build/O2/o2-dataflow-cs8`: Three QC tests `o2-qc-test-core`, `testTriggers`, and `testCheckWorkflow` all fail due to time-outs, unrelated to this MR - **False negative** 

Yours, 
_Christian_

**Comment by cholmcc** [comment]
Now also `build/O2/o2` and `build/O2/fullCI` fails due to timeout in unrelated QC test.  

Also, please remove `stale` label from this MR.   Thanks. 

Yours, 
_Christian_

**Comment by cholmcc** [comment]
Hi all, 

I made an [MR](https://github.com/alisw/arrow/pull/12) against ALICE's `arrow` fork to fix the problem with `build/AliceO2/O2/o2/macOS{,-arm}`.  I also opened a [bug report](https://github.com/apache/arrow/issues/38709) for upstream `arrow`. 

Yours,
_Christian_

**Comment by cholmcc** [comment]
Hi all,

Any news on this? 

Yours,
_Christian_

**Comment by cholmcc** [comment]
Please advice on next steps to get this merged.  Thanks.

**Comment by sawenzel** [comment]
Sorry for the delay. We are still quite busy with commissioning work for upcoming PbPb productions and currently fixing known bugs in AOD conversion (which might interfere here). I am afraid, given the amount of changes, a proper review might still take some time but it's definitely on my list.

You could do me a favour and clean up your PR. Currently, there are 88 commits and it would be good to rebase on dev and just force-push one single (or few) commit with the proposed changes.

**Comment by cholmcc** [comment]
Hi Sandro, 

>  I am afraid, given the amount of changes ...

The changes are not _that_ big, and are isolated to the MC part only.   That's why I did a thorough test on `o2-aod-mc-publisher-workflow`.  

> You could do me a favour and clean up your PR. Currently, there are 88 commits and it would be good to rebase on
> dev and just force-push one single (or few) commit with the proposed changes.

Could you give some rather detailed instructions on how to do that?  I'm a little worried that things will break.    Thanks. 

Yours, 
_Christian_

**Comment by cholmcc** [comment]
OK, I figured it out myself:

```
git reset --soft $(git merge-base dev HEAD)
git commit 
git push --force
```
For future reference (ignore for the discussion of the MR):

- `git reset --soft $(git merge-base dev HEAD)` resets working directory to last head of `dev`.  Changed files are stil staged for commit. 
- `git commit` commits all changes in working directory as a single commit.  Here, I used `git log` to get the commit message I originally used, and copy'n'pasted that 
- `git push --force` rewrites the branch history starting from `$(git merge-base dev HEAD)` and up to the last commit, so that it only contains a single commit. 

Note `git push --force` can be a bit dangerours.

**Comment by sawenzel** [comment]
@cholmcc : I now find some time to take a look into this. But, unfortunately, some merge conflicts need to be fixed first of all. Could you take care of this (and force-push a single commit)? Thx.

**Comment by cholmcc** [comment]
Hi Sandro, 

Sorry for the late reply - been a bit busy the past week or so. 

I will try to get around to fixing the merge conflicts.   Note sure if I can make it before Christmas though. 

Yours,

_Christian_

**Review by sawenzel — CHANGES_REQUESTED** [review]
Hi. I started studying the code in this PR. At the moment, more work is needed because
the PR does not actually compile.

Thereafter, please also perform a validation step by running a simple O2DPG MC workflow:
(alibuild build O2sim; alienv enter O2sim; .... then run a script like)
```
#!/bin/bash

#
# A example workflow MC->RECO->AOD for a simple pp production
# excluding ZDC

NWORKERS=${NWORKERS:-8}
SIMENGINE=${SIMENGINE:-TGeant4}

# create workflow
${O2DPG_ROOT}/MC/bin/o2dpg_sim_workflow.py -eCM 14000 -col pp -gen pythia8 -proc "cdiff" -tf 4  -ns 200 -e ${SIMENGINE} -j ${NWORKERS}                               \
                                                     -run 303000 -seed 624 -interactionRate 50000 

# run workflow
${O2DPG_ROOT}/MC/bin/o2_dpg_workflow_runner.py -f workflow.json -tt aod
```
I would say we can go ahead from my side once this succeeds and the AOD content is compatible with now.

**Comment by cholmcc** [comment]
Hi, 

I ran 

```
#!/bin/bash
#
# A example workflow MC->RECO->AOD for a simple pp production
# excluding ZDC
NWORKERS=${NWORKERS:-6}
SIMENGINE=${SIMENGINE:-TGeant4}

${O2DPG_ROOT}/MC/bin/o2dpg_sim_workflow.py \
             -eCM 14000 \
             -col pp \
             -gen pythia8 \
             -proc "cdiff" \
             -tf 2  \
             -ns 50 \
             -e ${SIMENGINE} \
             -j ${NWORKERS} \
             -run 303000 \
             -seed 624 \
             -interactionRate 50000 

${O2DPG_ROOT}/MC/bin/o2_dpg_workflow_runner.py -f workflow.json -tt aod
```

which however failed (though not in the AOD part which is relevant to this MR): 

```
Launching task: ${O2_ROOT}/bin/o2-sim-digitizer-workflow -b --run --condition-not-after 3385078236000 -n 50 --sims sgn_1 --interactionRate 50000 --incontext collisioncontext.root --disable-write-ini --configKeyValues "HBFUtils.orbitFirstSampled=0;HBFUtils.nHBFPerTF=128;HBFUtils.orbitFirst=0;HBFUtils.runNumber=303000;HBFUtils.startTime=1547590800000;MFTAlpideParam.roFrameLengthInBC=198;ITSAlpideParam.roFrameLengthInBC=198" --onlyDet ITS,TOF,FDD,MCH,MID,MFT,HMP,EMC,PHS,CPV --ccdb-tof-sa --combine-devices  &> restdigi_1.log &
...
command ${O2_ROOT}/bin/o2-sim-digitizer-workflow -b --run --condition-not-after 3385078236000 -n 50 --sims sgn_1 --interactionRate 50000 --incontext collisioncontext.root --disable-write-ini --configKeyValues "HBFUtils.orbitFirstSampled=0;HBFUtils.nHBFPerTF=128;HBFUtils.orbitFirst=0;HBFUtils.runNumber=303000;HBFUtils.startTime=1547590800000;MFTAlpideParam.roFrameLengthInBC=198;ITSAlpideParam.roFrameLengthInBC=198" --onlyDet ITS,TOF,FDD,MCH,MID,MFT,HMP,EMC,PHS,CPV --ccdb-tof-sa --combine-devices
restdigi_1 failed ... checking retry
```

What I see in `restdigi_1.log` is 

```
[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/37}
[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/38}
[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/39}
[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/40}
[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/41}
[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/42}
[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/43}
[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/44}
[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/45}
```

The point, however, is that none of these problems have anything to do with the MR in question. 

Yours,

_Christian_

File: Detectors/AOD/src/aod-producer-workflow.cxx [hunk 1]
```diff
@@ -36,7 +36,7 @@ void customize(std::vector<ConfigParamSpec>& workflowOptions)
     {"disable-root-output", o2::framework::VariantType::Bool, false, {"disable root-files output writer"}},
     {"disable-mc", o2::framework::VariantType::Bool, false, {"disable MC propagation"}},
     {"disable-secondary-vertices", o2::framework::VariantType::Bool, false, {"disable filling secondary vertices"}},
-    {"disable-strangeness-tracker", o2::framework::VariantType::Bool, false, {"disable filling strangeness tracking"}},
+    {"disable-strangeness-tracking", o2::framework::VariantType::Bool, false, {"disable filling strangeness tracking"}},
```
Comment (shahor02) [inline Detectors/AOD/src/aod-producer-workflow.cxx]: Please don't modify options, even if it sounds better it will break dependent scripts.

File: Detectors/AOD/src/aod-producer-workflow.cxx [hunk 2]
```diff
@@ -52,7 +52,7 @@ WorkflowSpec defineDataProcessing(ConfigContext const& configcontext)
   o2::conf::ConfigurableParam::updateFromString(configcontext.options().get<std::string>("configKeyValues"));
   auto useMC = !configcontext.options().get<bool>("disable-mc");
   bool enableSV = !configcontext.options().get<bool>("disable-secondary-vertices");
-  bool enableST = !configcontext.options().get<bool>("disable-strangeness-tracker");
+  bool enableST = !configcontext.options().get<bool>("disable-strangeness-tracking");
```
Comment (shahor02) [inline Detectors/AOD/src/aod-producer-workflow.cxx]: idem

**Review by shahor02 — CHANGES_REQUESTED** [review]

**Comment by cholmcc** [comment]
OK, so I think my accesment was wrong. What really was the problem was that the `restdigi` job didn't get `FT0/TRIGGERINPUT` and `FV0/TRIGGERINPUT`.   I "fixed" that by modifying the command line to 

```
${O2_ROOT}/bin/o2-sim-digitizer-workflow 
  -b \
  --run \
  --condition-not-after 3385078236000 \
  -n 50 \
  --sims sgn_1 \
  --interactionRate 50000 \
  --incontext collisioncontext.root \
  --disable-write-ini \
  --configKeyValues "HBFUtils.orbitFirstSampled=0;HBFUtils.nHBFPerTF=128;HBFUtils.orbitFirst=0;HBFUtils.runNumber=303000;HBFUtils.startTime=1547590800000;MFTAlpideParam.roFrameLengthInBC=198;ITSAlpideParam.roFrameLengthInBC=198" \
  --onlyDet ITS,TOF,FDD,MCH,MID,MFT,HMP,EMC,PHS,CPV,FT0,FV0 \
  --ccdb-tof-sa \
  --combine-devices
```

It seems that `Steer/DigitizerWorkflow/src/SimpleDigitizerWorkflow.cxx` should be amended to call `o2::ft0::getDigitReaderSpec` (from `Detectors/FIT/FT0/workflow/src/DigitReaderSpec.cxx`) in case the FT0 detector is not configured for output and triggered detectors are enabled - and the same for FV0 - otherwise the other detectors will not get the `FT0/TRIGGERINPUT` table - In particular EMCAL seems to have a need for these tables.  It can be turned off by the option `--no-require-ctpinputs-emc` but that is not passed by the DPG Python script(*).  I changed the script to do that, and removed the FT0 and FV0 from the `restdigi` command line, and that seems to fix the issue too. 

(*)I honestly didn't think it was possible to write that poorly structured Python code - ah well.

**Comment by cholmcc** [comment]
OK, yet another error - this time in call to `o2-secondary-vertexing-workflow`.  The options listed a lot of track combinations with TPC (`TPC-ITS`, ...) but never `TPC` alone.  This made the workflow fail immediately with a `FATAL` signal.  Again a small fix to the Python script seemed to fix the issue.

**Comment by shahor02** [comment]
> OK, yet another error - this time in call to `o2-secondary-vertexing-workflow`. The options listed a lot of track combinations with TPC (`TPC-ITS`, ...) but never `TPC` alone.

Are you using obsolete O2DPG? The requirement of TPC in the sources is recent and is accounted in the `o2dpg_sim_workflow.py`.

> (*)I honestly didn't think it was possible to write that poorly structured Python code - ah well.
better to avoid such comments...

**Comment by cholmcc** [comment]
```
Launching task:  [ -f aodmerge_input.txt ] && rm aodmerge_input.txt; for i in `seq 1 2`; do echo "tf${i}/AO2D.root" >> aodmerge_input.txt; done; o2-aod-merger --input aodmerge_input.txt --output AO2D.root &> aodmerge.log &

**** Pipeline done success (global_runtime : 6.399s) *****
```

Yay! It seemed to work.

Looking in the various trees I see data as expected, including the new HepMC auxiliary tables (the HeayIon table is clearly empty because it was a pp simulation).

**Comment by cholmcc** [comment]
I did 

```
o2-analysis-mm-rivet \
  --hepmc-recenter \
  --rivet-analysis  ALICE_YYYY_I1234567 \
  --rivet-pwd \
  --rivet-sources  ALICE_YYYY_I1234567.cc \
  --rivet-finalize --aod-file AO2D.root \
  --rivet-dump RivetResults.yoda \
  --batch 
```

(code from [Rivet for O2](https://github.com/cholmcc/O2Physics/tree/cholmcc_pwgmm_rivet) - in particular from [`PWGMM/Rivet`](https://github.com/cholmcc/O2Physics/tree/cholmcc_pwgmm_rivet/PWGMM/Rivet))

followed by 

```
./ALICE_YYYY_I1234567.py -s RivetResults.yoda 
```

or alternatively 

```
root -l ../src/O2Physics/PWGMM/Rivet/macros/GetRivetAOs.macro
./ALICE_YYYY_I1234567.py -s RivetResults.yoda 
```

to get the following plot 

RivetResults

I think the MR has lived up to its promise :smile:

**Comment by cholmcc** [comment]
> Are you using obsolete O2DPG? The requirement of TPC in the sources is recent and is accounted in the 
>  o2dpg_sim_workflow.py.

Yes, it was a little old - but not terribly. 

>> (*)I honestly didn't think it was possible to write that poorly structured Python code - ah well.
> better to avoid such comments...

None the less true - that code suffers from Fortrantitis in a bad way :smile:   Typically when I see Python code, even by novice programmers, it is OK structured and clear, which is why this code surprised me.  I guess it must be a pain to maintain. I understand how this comes about - I'll do quick fix here, then another fix here, and oh, its more important that it worked yesterday than it will work for eternity and I will come back and tidy it up - and before you know it, you have a mess on your hands.

**Comment by sawenzel** [comment]
@cholmcc : Do you believe that insulting people in commit messages and PR discussions is helpful?
If it continues like this, my motivation to review your developments and give support will just decrease.
I honestly believe there are more professional ways of discussing issues. Sometimes it is just enough to shortly ping about problems on mattermost and get hints of how to resolve them.

Part of your challenges come from the fact that you are not using the collaboration rules and the official workflow of using alidist for software distribution, which defines a tested and curated software portfolio. In there, O2DPG is fully up to date and guaranteed to run the example workflow that I gave to you (because we test it nightly). 

I suggested running the above workflow, because when I took time to review your developments:
a) your PR did not compile
b) even after fixing your PR privately, the AOD conversion step crashed

**Comment by cholmcc** [comment]
> @cholmcc : Do you believe that insulting people in commit messages and PR discussions is helpful?

I don't think I insulted anyone - if someone felt that way, well I'd say they have rather thin skin.  I certainly didn't point anyone out in particular, only that the code is truly horrible. 

What I do think is helpful is to point out that some code is really poorly written and it would make sense, not least for maintenance purposes, fix it.  

> If it continues like this, my motivation to review your developments and give support will just decrease.

Should that be considered a threat?   Because that is clearly not OK. 

Look, I did not do this work for my sake - I did it because I was tasked by the collaboration to do a job - namely integrate Rivet into O2.   This MR is just one piece of that puzzle.  So you are not doing me, but the collaboration, wrong if you chose to stop doing your job. 

> I honestly believe there are more professional ways of discussing issues. 

You mean like issuing veiled threats? 

> Sometimes it is just enough to shortly ping about problems on mattermost and get hints of how to resolve them.

I really don't want to solve the problems of the DPG scripts - those scripts are so obtuse that it would take a considerable effort to do so - in fact, I think it would most likely be more efficient to start over again on those scripts. 

Trust, me I've ping many people on Mattermost and else where how to do a full MC->AOD test.  It wasn't until you kindly gave the instruction that I knew how to do it.  When it then turns out that that didn't really work because of problems in the DPG scripts - well - then I got even more frustrated than before. 

> Part of your challenges come from the fact that you are not using the collaboration rules 

Which rules?  There's no hard and fast rule that you _need_ to use `alidist`.   

> ... and the official workflow of using alidist for software distribution, ...

An no, the problems does not stem from me not using `alidist` at all, but rather that the tools don't actually work as intended.   If you think that the problems arise from not using `alidst`, then I'm afraid there's something you have misunderstood. 

> .. which defines a tested and curated software portfolio.

I'm aware that not using `alidist` puts a burden on me to solve some issues on my own, and I routinely do so (much more than I care for - it is no fun digging through thousands of lines of code only to find that a certain assumption was made about the something in the environment which will only ever be true in very specific circumstances). 

As for "tested" - it seems the tests are basically - does it run on LXPLUS, OK, we're good.   That's not how one does testing of software.  Testing software involves meticulous attention to corner cases and presumptions.  Try to deploy `alidist` on some other platform than CentOS and you will find it rather burdensome.  

>  In there, O2DPG is fully up to date and guaranteed to run the example workflow that I gave to you (because we test it nightly).

OK, so there seems to be something going on in your test environment that does not comport with a vanilla environment.   F.ex., without the option `--no-require-ctpinputs-emc` the code _will_ crash - it's just a matter of looking at the sources to see that.  This is what I mean by "testing" is done in a very special environment.  It would be one thing if this environment was properly documented, but sadly it isn't. 

BTW, there's no guaranty - even with `alidist`, and especially when doing development - that the various tools are in sync.   It can be done somewhat in deployment targets, but even there it can break (vis the debacle over `fastjet`). 

> I suggested running the above workflow, because when I took time to review your developments:
> a) your PR did not compile

This was due to some other change somewhere else in O2.   That the `o2::framework::Output` dropped the `Lifetime` argument (commit #f6abfb04d5eedf284fd6913bf01820ed82af6a3b on the 6th of December by Guilio), so you cannot pin that on the PR.  Of course, had you looked at the PR earlier - say in November, it would not have been a problem, 

> b) even after fixing your PR privately, the AOD conversion step crashed

It would have been helpful if you could have reported the error you saw when you saw it.   I have no way of gauging what the problem might have been for you without that input (or rather output :smile:) - could be because of missing  `--no-require-ctpinputs-emc` or something entirely different.    After I fixed the various merge conflicts (not that many) between this MR and the latest `dev`, I did not see any more problem in the code but only had trouble with the DPG scripts. 

Anyways, I think we're at a point where this PR is fairly well tested, and I think it will make sense to merge it at this point.  Then, I will get out of your hair so you can get around to something that is more fun for you.   Thanks for all your help. 

Yours,

_Christian_

**Comment by pzhristov** [comment]
@cholmcc I am afraid you misunderstood what was the job description of Sandro. He is helping you not because this is his task, but because we want to move the things on. So this kind of discussion is not welcome in the PRs. If you have any question on this, I am happy to chat with you privately. At the end we have been working together for about 20 years :)

**Comment by pzhristov** [comment]
@cholmcc I propose to run the original code, then run again the AOD creator with these modifications on the same input and see what is the difference between the original AODs and the newly created ones. If we are happy with the result, we can merge.

**Comment by cholmcc** [comment]
> @cholmcc I propose to run the original code, then run again the AOD creator with these modifications on the same input and see what is the difference between the original AODs and the newly created ones. If we are happy with the result, we can merge.

That's a non-trivial thing to do, as you know.  I means switch to the `dev` branch and running with that.  See what I can do.

**Comment by cholmcc** [comment]
Hi all, 

I pushed a new singe commit to this MR.  

I found a few bugs after comparing to output from `dev`.  I won't go in to all the details - you can find them in the [commit message](https://github.com/AliceO2Group/AliceO2/pull/12039/commits/f87bad43db186765c24b22f1b53b6bbb51c30225). 

Here, I will show you the results of my comparisons between `dev` and this MR.  

To be transparent about what I did: 

- I ran script [posted above](https://github.com/AliceO2Group/AliceO2/pull/12039#issuecomment-1896552987) (including the fix to DPG to pass `----no-require-ctpinputs-emc` when digitising) for this MR. 
- I copied the generated `AO2D.root` to storage `mine/A02D.root`
- I did 
  
      rm -f tf?/aod_?.log_done aodmerge.log_done 

- Then I switch to the `dev` branch and ran the [posted script](https://github.com/AliceO2Group/AliceO2/pull/12039#issuecomment-1896552987)  again (again, thanks Sandro for providing this script).
- I copied the generated `AO2D.root` to storage `dev/A02D.root`
- Then I ran 
[this script](https://github.com/AliceO2Group/AliceO2/files/13987135/CompareAO2D.C.txt) (remove `.txt` from the file name) to get 
[this report](https://github.com/AliceO2Group/AliceO2/files/13987151/compare.pdf)

What I see is that all comes out the same between `dev` and this MR - _except_ `trackqa`.  

I spent quite a bit of time on the `trackqa` discrepancies.   The important thing to keep in mind is, as far as I can tell, that the filling of the `trackqa` tables does not, _in any way_, depend on MC data.  This MR _solely_ effects the MC part (`mc...`), and the data (reco) part is _entire_ unchanged.  My best guess is that the QA part has some randomness built into it. I believe I say different differences between runs, which would comport with that.   If you have some other ideas, please let me know.  Thanks. 

I also did a closure test of the event before and after AOD generation.  That is, I took the kinematics tree from one timeframe of the simulation and generated a `kine_AO2D.root` file from that 

    o2-sim-kine-publisher --kineFileName sgn_1 | o2-sim-mctracks-to-aod 
    mv AnalysisResults_trees.root kine_AO2D.root 

Then, from both this `kine_AO2D.root` and the `AO2D.root` files above, I generated HepMC files 

    for part in dev mine kine : do 
        o2-aod-mc-to-hepmc --hepmc-dump ${part}/AO2D.kine --aod-file ${part}/AO2D.root ; done 

and then illustrated a few events with 

    $O2_PHYSICS_SOURCE/PWGMM/Rivet/Tools/hepmc.py 

from my [O2Physics Rivet branch](https://github.com/cholmcc/O2Physics/tree/cholmcc_pwgmm_rivet).  I attach them here in a [zip archive](https://github.com/AliceO2Group/AliceO2/files/13987550/events.zip).   Note that the `kine` illustrations will have more particles than the AOD ones, because `kine` contains the full event, including particles created during simulation transport, while the AOD ones only has those stored in the AODs.   What we can learn from this is that the event structure is preserved from simulation output to the AOD output - not a bad thing :smile: 

Of course, I would be useful to also illustrate the event from the output of the EG.  However, that data isn't explicitly stored by the DPG script because Pythia is run as an internal thing in the simulations.   However, I've done that comparison before (EG and post simulation) and that checked out too. 

So the chain from EG->SIM->DIGIT->RECO->AOD seems to preserve the event structure - also with this MR :smile: 

Yours,

_Christian_

**Comment by sawenzel** [comment]
@cholmcc : Thanks for doing this validation. As you say, the important part is the agreement of the MC tables. Track tables may indeed vary a bit since some digitisation algos and TPC reconstruction has random number usage with multi-threading that is not completely reproducible between runs.

Your script may be useful for doing this validation in more automatic ways in the future.

To me all looks good now and if there are no further objections, the PR will be merged end of the day.

**Comment by cholmcc** [comment]
Hi Sandro, 

> @cholmcc : Thanks for doing this validation. 

It was _not_ (**edit**) as much work as I first thought it to be :smile:

> Track tables may indeed vary a bit since some digitisation algos and TPC reconstruction has random number usage 
> with multi-threading that is not completely reproducible between runs. 

OK, so I guess its likely that randomness is part of the reason why the `trackqa` tables do not match exactly.  Great.

> Your script may be useful for doing this validation in more automatic ways in the future.

By all means take it and put it somewhere accessible. 

> To me all looks good now and if there are no further objections, the PR will be merged end of the day.

That's great.  Looking forward to that.  

Could you have a peek at the issue I raised in the [commit log](https://github.com/AliceO2Group/AliceO2/pull/12039/commits/f87bad43db186765c24b22f1b53b6bbb51c30225) - namely with particle selection I believe is currently broken in `dev` and in this MR.   I think, if we agree that there's a problem in `dev`, that the best approach is to take this MR as-is, and then I do a simple PR to fix that selection.   I think it creates the best most-tractable history. 

Thanks again. 

Yours,

_Christian_

**Comment by cholmcc** [comment]
`Formatting / PR formatting / clang-format (pull_request_target)` seems to fail for some weird reason - it cannot find the `dev` branch, whatever that means.  I did check all the code for formatting, and all was good. I would classify it as a **false negative**.

**Comment by cholmcc** [comment]
Thank you for merging this.   We are now one step closer to getting **Rivet** fully integrated into $O^2$.  With this in  $O^2$ an analyser can now use official MC productions to test out their implementation of **Rivet** analyses, and use those analyses for model comparison when preparing manuscripts for publication.  The other steps missing are getting **Rivet** [properly deployed](https://github.com/alisw/alidist/pull/5320) in the various computation environments and getting the [**Rivet** wrapper](https://github.com/cholmcc/O2Physics/tree/cholmcc_pwgmm_rivet/PWGMM/Rivet) accepted into $O^2\mathrm{Physics}$. 

Thank you. 

Yours,

_Christian_
